tap_gitlab/client.py

```diff
@@ -144,13 +145,53 @@ def post_process(self, row: dict, context: Optional[dict] = None) -> Optional[di
 class ProjectBasedStream(GitLabStream):
     """Base class for streams that are keys based on project ID."""
 
-    state_partitioning_keys = ["project_path"]
+    state_partitioning_keys = ["project_id"]
+
+
+class NoSinceProjectBasedStream(ProjectBasedStream):
+    """Base class for streams lacking a "since" query parameter.
+
+    It includes logic to emulate a "since" query parameter, for
+    instance for notes streams (eg. issue notes, MR notes...).
+    """
+
+    def get_next_page_token(
+        self, response: requests.Response, previous_token: Optional[Any]
+    ) -> Optional[Any]:
+        """Emulate a "since" parameter for streams that do not support it.
+
+        Return a token for identifying next page or None if no more pages.
+        """
+        # extract the cutoff time from request parameters
+        request_parameters = parse_qs(str(urlparse(response.request.url).query))
+        # parse_qs interprets "+" as a space, revert this to keep an aware datetime
+        try:
+            cutoff = (
+                request_parameters[self.bookmark_param_name][0].replace(" ", "+")
+                if self.bookmark_param_name in request_parameters
+                else None
+            )
+        except IndexError:
+            cutoff = None
+
+        if cutoff is not None:
+            # get result items from response
+            resp_json = response.json()
+            results = resp_json
+            if len(results) == 0:
+                # gitlab API sometimes returns empty lists, no need to paginate further
+                return None
+            # if we receive items past the cutoff timestamp, we can stop paginating
+            if parse(results[-1][self.replication_key]) < parse(cutoff):
+                return None
+
+        return super().get_next_page_token(response, previous_token)
 
 
 class GroupBasedStream(GitLabStream):
     """Base class for streams that are keys based on group ID."""
 
-    state_partitioning_keys = ["group_path"]
+    state_partitioning_keys = ["group_id"]
```

Comment:
My reasoning was strongly influenced by what we've seen in [this tap-github issue](https://github.com/MeltanoLabs/tap-github/issues/110).
Paths can be renamed, potentially leading to inconsistencies, whereas `id`s should remain stable across renames. Depending on how the tap is used (which `group` values are passed in config) passing `newOrgName` in config while `oldOrgName` is in the state from a previous run would cause the tap to miss the bookmarks entirely, and potentially redownload a lot of data. Using `group_id` should prevent such problems.

The result is indeed a bit less readable than using the `group_path`.

Happy to hear your thoughts on this.

Reply:
@laurentS - I had not thought about renames! Very good argument - thank you. I'm inclined to agree with this change, although I think it should be considered a breaking change.

I'll open a new issue to consolidate and strategize around breaking changes like this one.

Reply:
See my note below. I just added this, as I think it makes sense. The tap should now run on numeric ids instead of paths, both for `projects` and `groups`. Casing is corrected and invalid names are removed from the list of partitions.